csrc/dd_arith.c

```diff
@@ -488,3 +494,244 @@ ddouble tanhq(ddouble a)
     c = sqrtq(adddq(1.0, sqrq(s)));
     return divqq(s, c);
 }
+
+ddouble tanq(ddouble a)
+{
+    if (iszeroq(a))
+        return Q_ZERO;
+
+    ddouble s, c;
+    s = sinq(a);
+    c = cosq(a);
+    return divqq(s, c);
+}
+
+void sincos(const ddouble a, ddouble *sin_a, ddouble *cos_a) {
+
+    if (iszeroq(a)) {
+        *sin_a = Q_ZERO;
+        *cos_a = Q_ONE;
+        return;
+    }
+
+    // approximately reduce modulo 2*pi
+    ddouble z = nintq(divqq(a, Q_2PI));
+    ddouble r = subqq(a, mulqq(Q_2PI, z));
+
+    // approximately reduce module pi/2 and pi/16
+    ddouble t;
+    double q = floor(r.hi / Q_PI_2.hi + 0.5);
+    t = subqq(r, mulqd(Q_PI_2, q));
+    int j = (int)(q);
+    int abs_j = (int)(fabs(j));
+    q = floor(t.hi / Q_PI_16.hi + 0.5);
+    t = subqq(t, mulqd(Q_PI_16, q));
+    int k = (int)(q);
+    int abs_k = (int)(fabs(k));
```

Comment:
I think this should just be `abs`, since `k` is integer? Or is there something wrong with the logic?

Reply:
It definitely should be `abs` instead of `fabs`. Oops! This has been fixed in the recent commits.